modules/fs2/src/main/scala/com/snowplowanalytics/snowplow/enrich/fs2/Payload.scala

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2020 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+package com.snowplowanalytics.snowplow.enrich.fs2
+
+import scala.annotation.tailrec
+
+import cats.Applicative
+import cats.syntax.either._
+import cats.data.Validated
+
+import fs2.{Pure, Stream}
+
+import com.snowplowanalytics.snowplow.enrich.fs2.Payload.Parsed
+
+/**
+ * Anything that has been read from [[RawSource]] and needs to be acknowledged
+ * or a derivative (parsed `A`) that can be used to acknowledge the original message
+ * @param data original data or anything it has been transformed to
+ * @param ack a side-effect to acknowledge the message or no-op in case the original
+ *            message has been flattened into multiple rows and only last row contains
+ *            the actual side-effect
+ */
+case class Payload[F[_], A](data: A, ack: F[Unit]) {
+
+  /**
+   * Flatten all payloads from a list and replace an `ack` action to no-op everywhere
+   * except last message, so that original collector payload (with multiple events)
+   * will be ack'ed only when last event has sunk into good or bad sink
+   */
+  def decompose[L, R](implicit ev: A <:< List[Validated[L, R]], F: Applicative[F]): Stream[F, Parsed[F, L, R]] = {
+    val _ = ev
```

Comment:
What does this line do?

Reply:
Just "uses" `ev`. Otherwise scalac compiles that there's an argument never used.

Reply:
If it's not used apart from this line, why does the implicit arg need to be there?

Reply:
It's an compile-time evidence that our `A` is `List[Validated[L, R]]`, so we can use `mapWithLast` on it, which works only with `List`. And our `use` works only with `Validated[L, R]`.

In other words, if `A` is something else, e.g. `Array[Byte]` you won't be able to call `decompose` - it won't compile.

Reply:
I see, thanks for the explanation! So in `Payload[A]` `A` can be anything but if we want to use `decompose` A needs to be a `List[Validated[L, R]`, is that right?

Weird that Scala forces us to write `val _ =`

Reply:
```
case class Wapper[A](a: A) {
  /** Exists only if its an Int */
  def square(implicit ev: A <:< Int): Int =
    a * a

  def capitalize(implicit ev: A <:< String): String =
    a.capitalize
}
```

Reply:
>  So in Payload[A] A can be anything but if we want to use decompose A needs to be a List[Validated[L, R], is that right?

That's right, yeah.

> Weird that Scala forces us to write val _ =

It's just a value in the end. Moreover we still can use it - it has `apply[A](a: A): List[Validated[L, R]]` method. But I noticed that people prefer `asInstanceOf` in this case.